File: 5331/project4/banker.cpp
//*********************************************************
//
// Shuyi Zhao
// Operating Systems
// Project #4: Implementation of Banker's Algorithm
// April 13, 2011
//
//**********************************************************
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <iomanip>

using namespace std;

void insertNum(int& n, int& m, ifstream& myInFile);
void insertV(int* array, int m, ifstream& myInfile);
void insertReq(int* requestV, int m, ifstream& myInFile);
void insertM(int** array, int n, int m, ifstream& myInfile);
void displayNum(int n, int m);
void displayV(int* array, int m);
void displayM(int** array, int n, int m);
void foundNeed(int** claimM, int** allocationM, int** needM, int n, int m);
bool isSafeState(int n, int m, int* available, int** needM);
bool isRequestGranted(int* requestV, int* availableV, int** needM, int** allocationM, int m);

int main(int argc, char* argv[])
{
    ifstream myInFile;
       
    int n;//number of processes
    int m;//number of resource types
    
    //fileName=argv[1];
    myInFile.open(argv[1]); 
    insertNum(n, m, myInFile);

	int* resourceV=new int[m];
    int* availableV=new int[m];
    int* requestV=new int[m+1];
    int** claimM=new int*[n];
    for(int i=0; i<n; i++)
    {
        claimM[i]=new int[m];
    }
    int** allocationM=new int*[n];
    for(int i=0; i<n; i++)
    {
        allocationM[i]=new int[m];
    } 
    int** needM=new int*[n];
    for(int i = 0; i<n; i++)
    {
        needM[i]=new int[m];
    }

    insertV(resourceV, m, myInFile);
    insertV(availableV, m, myInFile);
    insertM(claimM, n, m, myInFile);
    insertM(allocationM, n, m, myInFile);
    insertReq(requestV, m, myInFile);

    
    displayNum(n, m);
    cout << "The Resource Vector is..." << endl;
    displayV(resourceV, m);
    cout << "The Available Vector is..." << endl;
    displayV(availableV, m);
	cout << "The Claim Matrix is..." << endl;
    displayM(claimM, n, m);
	cout << "The Allocation Matrix is..." << endl;
    displayM(allocationM, n, m);
	cout << "The Need Matrix is..." << endl;
    foundNeed(claimM, allocationM, needM, n, m);
    displayM(needM, n, m);
    
    if(isSafeState(n, m, availableV, needM))
    {
		cout << "THE SYSTEM IS IN A SAFE STATE" << endl << endl;
        if(isRequestGranted(requestV, availableV, needM, allocationM, m))
            {
                if(isSafeState(n, m, availableV, needM))
                {
                    cout << "THE REQUEST CAN BE GRANTED: NEW STATE FOLLOWS" <<endl;
					cout << "The Resource Vector is..." << endl;
                    displayV(resourceV, m);
					cout << "The Available Vector is..." << endl;
                    displayV(availableV, m);
					cout << "The Claim Matrix is..." << endl;
                    displayM(claimM, n, m);
					cout << "The Allocation Matrix is..." << endl;
                    displayM(allocationM, n, m);
					cout << "The Need Matrix is..." << endl;
                    displayM(needM, n, m);
                }
                else
                    cout << "THE REQUEST CANNOT BE GRANTED" << endl;
            }
            else
            cout << "THE REQUEST CANNOT BE GRANTED" << endl;    
    }
    else
        cout << "THE SYSTEM IS NOT IN A SAFE STATE" << endl;
    
    
    delete[] resourceV;
    delete[] availableV;
    delete[] requestV;
    for(int i=0;i <n;i++) 
        delete[]   claimM[i]; 
    delete[]   claimM; 
    for(int i=0; i <n; i++) 
        delete[]   allocationM[i]; 
    delete[]   allocationM; 
    for(int i=0; i <n; i++) 
        delete[]   needM[i]; 
    delete[]   needM; 
    
    return 0;
}

//************************************************************
//
// Insert Number Function
//
// This function read the numbers of processes and types from file into parameters 
// 
//
// Return Value
// ------------
// void
//
// Reference Parameters
// --------------------
// int          n                      the number of processes
// int          m                      the number of types
// ifstream  myInFile                  streamReader
// Local Variables
// ---------------
// NONE
//
//*******************************************************************
void insertNum(int& n, int& m, ifstream& myInFile)
{
    myInFile >> n;
    myInFile >> m;
}

//************************************************************
//
// Insert Vector Function
//
// This function read data from file into array 
// 
//
// Return Value
// ------------
// void
//
// Reference Parameters
// --------------------
// int          *array                 the pointer point to a array
// int          m                      the number of types
// ifstream  myInFile                  streamReader
// Local Variables
// ---------------
// int          i
//
//*******************************************************************
void insertV(int* array, int m, ifstream& myInFile)
{
    for(int i=0; i<m; i++)
    {
        myInFile >> array[i]; 
    } 
}

//************************************************************
//
// Insert requestVector Function
//
// This function read data from file into request array 
// 
//
// Return Value
// ------------
// void
//
// Reference Parameters
// --------------------
// int          *requestV              the pointer point to request array
// int          m                      the number of types
// ifstream  myInFile                  streamReader
// Local Variables
// ---------------
// int          i
//
//*******************************************************************
void insertReq(int* requestV, int m, ifstream& myInFile)
{
    char opr;
    myInFile >> requestV[0];
    myInFile >> opr;
    for(int i=1; i<m+1; i++)
        myInFile >> requestV[i];  
}

//************************************************************
//
// Insert Matrix Function
//
// This function read data from file into 2D array 
// 
//
// Return Value
// ------------
// void
//
// Reference Parameters
// --------------------
// int          **array                the pointer point to a array which stores pointers point to array
// int          n                      the number of processes
// int          m                      the number of types
// ifstream  myInFile                  streamReader
// Local Variables
// ---------------
// int          i
// int          j
//
//*******************************************************************
void insertM(int** array, int n, int m, ifstream& myInFile)
{

     for(int i=0; i<n; i++)
     {
         for(int j=0; j<m; j++)
         {
             myInFile >> array[i][j]; 
         }
     }       
}

//************************************************************
//
// Display Number Function
//
// This function display the numbers of processes and types 
// 
//
// Return Value
// ------------
// void
//
// Reference Parameters
// --------------------
// int          n                      the number of processes
// int          m                      the number of types
// Local Variables
// ---------------
// NONE
//
//*******************************************************************
void displayNum(int n, int m)
{
    cout << "There are " << n << " processes and " << m << " resource types in the system." << endl << endl;
}

//************************************************************
//
// Display Vector Function
//
// This function display data from a array 
// 
//
// Return Value
// ------------
// void
//
// Reference Parameters
// --------------------
// int          *array                 the pointer point to a array
// int          m                      the number of types
// Local Variables
// ---------------
// char         a
// int          i
//
//*******************************************************************
void displayV(int* array, int m)
{
     for(char a='A'; a<'A'+m; a++ )
         cout << setw(2) << a;
     cout << endl;
     for(int i=0; i<m; i++)
         cout << setw(2) << array[i];
	 cout << endl << endl;
}

//************************************************************
//
// Display Matrix Function
//
// This function display data from 2D array 
// 
//
// Return Value
// ------------
// void
//
// Reference Parameters
// --------------------
// int          **array                the pointer point to a array which stores pointers point to array
// int          n                      the number of processes
// int          m                      the number of types
// Local Variables
// ---------------
// char         a
// int          i
// int          j
//
//*******************************************************************
void displayM(int** array, int n, int m)
{
     cout << "  ";
     for(char a='A'; a<'A'+m; a++ )
         cout << setw(2) << a;
     cout << endl;
     for(int i=0; i<n; i++)
     {
         cout << i << ":";
         for(int j=0; j<m; j++)
             cout << setw(2) << array[i][j];
		 cout << endl;
     }
	 cout << endl;
}

//************************************************************
//
// Find need matrix Function
//
// This function read data from claim matrix and allocation matrix to calculate the need matrix 
// 
//
// Return Value
// ------------
// void
//
// Reference Parameters
// --------------------
// int          **claimM               the pointer point to clainM[] which stores pointers point to claimM[][]
// int          **allocationM          the pointer point to allocationM[] which stores pointers point to allocationM[][]
// int          **needM                the pointer point to needM[] which stores pointers point to needM[][]
// int          n                      the number of processes
// int          m                      the number of types
// Local Variables
// ---------------
// int          i
// int          j
//
//*******************************************************************
void foundNeed(int** claimM, int** allocationM, int** needM, int n, int m)
{
     for(int i=0; i<n; i++)
         for(int j=0; j<m; j++)
             needM[i][j]=claimM[i][j] - allocationM[i][j];
	 
}

//********************************************************************
//
// Is Safe State Function
//
// This function test if the state is safe, if is, return true, 
// otherwise, return false.
//
// Return Value
// ------------
// bool                         True/False if state is safe
//
// Value Parameters
// ----------------
// int          *availableV            the pointer point to available array
// int          **needM                the pointer point to needM[] which stores pointers point to needM[][]
// int          n                      the number of processes
// int          m                      the number of types
//
// Local Variables
// ---------------
// int           *workV         the pointer point to work array
// bool          *Finish        the pointer point to finish array which stores the finish state of a process
// bool          changedFinish  hold the state if the process finish state has been changed in a loop
// int           counter        hold the number of elements in needM satisfied the conditions
// int           i
// int           j
// int           a
// int           b
//*******************************************************************
bool isSafeState(int n, int m, int* availableV, int** needM)
{
    //initialize step 1
    int* workV = new int[m];
    for(int i=0; i<m; i++)
        workV[i]=availableV[i];
    bool *Finish = new bool[n];
    for(int j=0; j<n; j++)
        Finish[j]=false;
    int counter;
    //bool safeState;
     
    //step2 step3
    bool changedFinish=true;
    while(changedFinish)
    {
		changedFinish=false;
        for(int a=0; a<n; a++)
        {
            if(Finish[a] == false)
            {
                counter=0;
                for(int b=0; b<m; b++)
                {
                    if(needM[a][b] <= workV[b])
                        counter++;                    
                }
                if(counter == m)
                {
                    for(int c=0; c<m; c++)
                    {
                        workV[c]=workV[c]+availableV[c];
                    }
                    Finish[a]=true;
                    changedFinish=true;
                }//end if(counter==m)                
            }//end if(Finish[a]==false)
        }//end for
    }  
    
    for(int d=0; d<n; d++)
    {
        if(Finish[d] == false)
            return false;
    }
    return true;
}

//********************************************************************
//
// Is Request Granted Function
//
// This function test is the new require vector is acceptable, if is
// display new state after accept vecotr, otherwise, display not granted 
//
// Return Value
// ------------
// bool                       
//
// Reference Parameters
// --------------------
// int          *availableV            the pointer point to available array
// int          **needM                the pointer point to needM[] which stores pointers point to needM[][]
// int          **allocationM          the pointer point to allocationM[] which stores pointers point to allocationM[][]
// int          **needM                the pointer point to needM[] which stores pointers point to needM[][]
// int          m                      the number of types
//
// Local Variables
// ---------------
// int		processNum	               read the requst process number from requestV
// int      a
// int      b
// int      c
//
//*******************************************************************
bool isRequestGranted(int* requestV, int* availableV, int** needM, int** allocationM, int m)
{
    int processNum=requestV[0];
    cout << "The Request Vector is..." << endl;
    cout << "  ";
    for(char c='A'; c<'A'+m; c++ )
        cout << setw(2) << c;
    cout << endl;
    cout << requestV[0] << ":";
    for(int b=1; b<m+1; b++)
        cout << setw(2) << requestV[b];
    cout << endl << endl;
    
    for(int a=0; a<m; a++)
    {
        if(requestV[a+1] <= needM[processNum][a])
        {
            if(requestV[a+1] <= availableV[a])
            {
                availableV[a]=availableV[a]-requestV[a+1];
                allocationM[processNum][a]=allocationM[processNum][a]+requestV[a+1];
                needM[processNum][a]=needM[processNum][a]-requestV[a+1];                
            }
            else
                return false;
        }
        else
            return false;
    } 
}

File: 5331/TEST/banker.cs
//********************************************************************
//
// Yuzhe Wei
// Operating Systems
// Programming Project #4: Banker's Algorithm
// Nov 23, 2010
// Instructor: Dr. Michael C. Scherger
//
//********************************************************************
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.IO;

namespace ConsoleApplication1
{
    public struct InputInfo     //struct store info from file 
    {
        public int process_num;
        public int resource_num;
        public int[] res_vec;
        public int[] avil_vec;
        public int[,] claim_Mrx;
        public int[,] allo_Mrx;
        public int pro_num;
        public int[] req_vec;
        public int[,] needMrx;
    }

    class Program
    {
        static void Main(string[] args)
        {
            string fileName;
            fileName = Console.ReadLine();
            StreamReader sr = File.OpenText(fileName);
            InputInfo Info = new InputInfo();
            readFile(ref sr, ref Info);
            Console.WriteLine("There are {0} processes and {1} resource types in the system.", Info.process_num, Info.resource_num);
            Console.WriteLine();
            Console.WriteLine("The resource Vector is ...");
            displayInfo(Info);
            Console.WriteLine();
            if (testState(Info))
            {
                Console.WriteLine("THE SYSTEM IS IN A SAFE STATE");
                allocRes(ref Info);
            }
            else
                Console.WriteLine("THE SYSTEM IS NOT IN A SAFE STATE");
                
            sr.Close(); //close reader
            
            Console.ReadKey();
        }

        //********************************************************************
        //
        // Read File Function
        //
        // This function read numbers from file and store them in the struct 
        // 
        //
        // Return Value
        // ------------
        // None
        //
        // Reference Parameters
        // --------------------
        // sr     StreamReader              Reader
        // Info   Inputinfo                 Struct hold file information
        // Local Variables
        // ---------------
        // temp		string	           temporary stroe one line in the file
        // m	    string		       temporary store one character
        //
        //*******************************************************************

        public static void readFile(ref StreamReader sr, ref InputInfo Info)
        {
            string temp;
            //read number of processes and resources
            temp = sr.ReadLine();
            string m;
            m = temp[0].ToString();
            Info.process_num = Int32.Parse(m);
            m = temp[2].ToString();
            Info.resource_num = Int32.Parse(m);

            int r = Convert.ToInt32(Info.resource_num);
            int p = Convert.ToInt32(Info.process_num);
            Info.res_vec = new int[r];
            Info.avil_vec = new int[r];
            Info.claim_Mrx = new int[p, r];
            Info.allo_Mrx = new int[p, r];
            Info.req_vec = new int[r];
            Info.needMrx = new int[p, r];


            //read 1*M resource  vector
            sr.ReadLine();
            temp = sr.ReadLine();
            temp = temp.Replace(" ", ""); //take out space from string

            for (int i = 0; i < temp.Length; i++)
            {
                int intTemp = Convert.ToInt32(temp[i]) - 48;
                Info.res_vec[i] = intTemp;

            }

            //read 1*M abaiable vector 
            sr.ReadLine();
            temp = sr.ReadLine();

            temp = temp.Replace(" ", ""); //take out space from string

            for (int i = 0; i < temp.Length; i++)
            {
                int intTemp = Convert.ToInt32(temp[i]) - 48;
                Info.avil_vec[i] = intTemp;

            }

            //read N*M claim Matrix       

            sr.ReadLine();
            for (int i = 0; i < p; i++)
            {
                temp = sr.ReadLine();
                temp = temp.Replace(" ", "");

                for (int j = 0; j < r; j++)
                {
                    int intTemp = Convert.ToInt32(temp[j]) - 48;
                    Info.claim_Mrx[i, j] = intTemp;
                }
            }

            //read a N*M allocation vector
            sr.ReadLine();
            for (int i = 0; i < p; i++)
            {
                temp = sr.ReadLine();
                temp = temp.Replace(" ", "");

                for (int j = 0; j < r; j++)
                {
                    int intTemp = Convert.ToInt32(temp[j]) - 48;
                    Info.allo_Mrx[i, j] = intTemp;
                }

            }

            //read i: 1*m request vector
            sr.ReadLine();
            temp = sr.ReadLine();

            Info.pro_num = temp[0] - 48;

            temp = temp.Replace(" ", "");
            temp = temp.Replace(":", "");


            for (int i = 1; i < temp.Length; i++)
            {
                int intTemp = Convert.ToInt32(temp[i]) - 48;
                Info.req_vec[i - 1] = intTemp;

            }

            //need matrix
            for (int i = 0; i < p; i++)
            {
                for (int j = 0; j < r; j++)
                {
                    Info.needMrx[i, j] = Info.claim_Mrx[i, j] - Info.allo_Mrx[i, j];
                }

            }

        }

        public static void displayInfo(InputInfo Info)
        {
            int r = Info.resource_num;
            char A = 'A';
           

            Console.Write("   ");
            for (int i = 0; i < r; i++)
            {
                Console.Write(A++);
                Console.Write(" ");
            }

            Console.WriteLine();
            Console.Write("   ");
            for (int i = 0; i < r; i++)
            {
                Console.Write(Info.res_vec[i]);
                Console.Write(" ");
            }
            Console.WriteLine();
            Console.WriteLine();
            //display avialable matrix
            Console.WriteLine("The resource Available Matrix is ...");
            Console.Write("   ");
            A = 'A';
            for (int i = 0; i < r; i++)
            {
                Console.Write(A++);
                Console.Write(" ");
            }
            Console.WriteLine();
            Console.Write("   ");
            for (int i = 0; i < r; i++)
            {
                Console.Write(Info.avil_vec[i]);
                Console.Write(" ");
            }
            Console.WriteLine();
            Console.WriteLine();
            //display claim matrix
            Console.WriteLine("The Claim Matrix is ...");
            Console.Write("   ");
            A = 'A';
            for (int i = 0; i < r; i++)
            {
                Console.Write(A++);
                Console.Write(" ");
            }
            Console.WriteLine();
            for (int i = 0; i < Info.process_num; i++)
            {
                Console.Write("{0}: ", i);

                for (int j = 0; j < r; j++)
                {

                    Console.Write(Info.claim_Mrx[i, j]);
                    Console.Write(" ");
                }
                Console.WriteLine();
            }


            Console.WriteLine();
            //display allocation matrix
            Console.WriteLine("The Allocation Matrix is ...");
            Console.Write("   ");
            A = 'A';
            for (int i = 0; i < r; i++)
            {
                Console.Write(A++);
                Console.Write(" ");
            }
            Console.WriteLine();
            for (int i = 0; i < Info.process_num; i++)
            {
                Console.Write("{0}: ", i);

                for (int j = 0; j < r; j++)
                {

                    Console.Write(Info.allo_Mrx[i, j]);
                    Console.Write(" ");
                }
                Console.WriteLine();
            }


            Console.WriteLine();

            //display need matrix
            Console.WriteLine("The Need Matrix is ...");
            Console.Write("   ");
            A = 'A';
            for (int i = 0; i < r; i++)
            {
                Console.Write(A++);
                Console.Write(" ");
            }
            Console.WriteLine();
            for (int i = 0; i < Info.process_num; i++)
            {
                Console.Write("{0}: ", i);

                for (int j = 0; j < r; j++)
                {

                    Console.Write(Info.needMrx[i, j]);
                    Console.Write(" ");
                }
                Console.WriteLine();
            }
        }

        //********************************************************************
        //
        // Test State Function
        //
        // This function test if the state is safe, if is, return true, 
        // otherwise, return false.
        //
        // Return Value
        // ------------
        // bool                         True/False if state is safe
        //
        // Value Parameters
        // ----------------
        // Info  	  InputInfo	        struct stored information of file
        //
        // Local Variables
        // ---------------
        // state		 bool		    the return value
        // process_count int            process counter
        // temp          int[]          hold one safe state process array
        //*******************************************************************
        public static bool testState(InputInfo Info)
        {
            bool state = false;
            bool[] finalState = new bool[Info.process_num];
            int process_count = 0;
            int[] temp = new int[Info.process_num];

            for (int i = 0; i < Info.resource_num; i++)
            {
                temp[i] = Info.avil_vec[i];
            }

            for (int i = 0; i < Info.process_num; i++)
            {
                finalState[i] = false;
            }

            for (int a = 0; a < Info.process_num; a++)
            {
                for (int i = 0; i < Info.process_num; i++)
                {
                    for (int j = 0; j < Info.resource_num; j++)
                    {
                        if (Info.needMrx[i, j] <= temp[j] && finalState[i] == false)
                            process_count++;

                        if (process_count == Info.resource_num)
                        {
                            finalState[i] = true;
                            for (int k = 0; k < Info.resource_num; k++)
                            {
                                temp[k] = temp[k] + Info.allo_Mrx[i, k];
                            }
                        }
                    }
                    process_count = 0;//reset process counter
                }
            }
            //check state
            for (int i = 0; i < Info.process_num; i++)
            {
                if (finalState[i] == true)
                    state = true;
                else
                    return false;
            }
            return state;

        }


        //********************************************************************
        //
        // Allocation Resource Function
        //
        // This function test is the new require vector is acceptable, if is
        // display new state after accept vecotr, otherwise, display not granted 
        //
        // Return Value
        // ------------
        // None                       
        //
        // Reference Parameters
        // --------------------
        // Info     InputInfo             struct stored information of file
        //
        // Local Variables
        // ---------------
        // A		char	 format for display: different resource
        //
        //*******************************************************************
        public static void allocRes(ref InputInfo Info)
        {
           
            char A;
            Console.WriteLine();
            Console.WriteLine("The request vector is ...");
            Console.Write("   ");
            A = 'A';
            for (int i = 0; i < Info.resource_num; i++)
            {
                Console.Write(A++);
                Console.Write(" ");
            }
            Console.WriteLine();
            Console.Write("{0}: ",Info.pro_num);
            for (int i = 0; i < Info.resource_num; i++)
            {
                Console.Write(Info.req_vec[i]);
                Console.Write(" ");
            }
            Console.WriteLine();
        
            for(int i = 0; i< Info.resource_num;i++)
                Info.avil_vec[i] = Info.avil_vec[i]-Info.req_vec[i];

            for (int i = 0; i < Info.resource_num; i++)
                Info.allo_Mrx[Info.pro_num, i] = Info.allo_Mrx[Info.pro_num, i]+Info.req_vec[i];

            for (int i = 0; i < Info.process_num; i++)
            {
                for (int j = 0; j < Info.resource_num; j++)
                {
                    Info.needMrx[i, j] = Info.claim_Mrx[i, j] - Info.allo_Mrx[i, j];
                }

            }
            
            if (testState(Info))
            {
                Console.WriteLine("THE REQUEST CAN BE GRANTED: NEW STATE FOLLOWS");
                displayInfo(Info);
            }
            else
            {
                Console.WriteLine("THE REQUEST CANNOT BE GRANTED");
                            
            }
        }

    }
   
}









